stdlib/database/sql/src/main/java/org/ballerinalang/sql/datasource/SQLDatasource.java

```diff
@@ -0,0 +1,241 @@
+                    sqlDatasourceToBeReturned = hikariDatasourceMap.compute(poolKey,
+                            (key, value) -> createAndInitDatasource(sqlDatasourceParams));
+                }
+            } finally {
+                existingSqlDatasource.releaseMutex();
+            }
+        } else {
+            sqlDatasourceToBeReturned = hikariDatasourceMap.computeIfAbsent(poolKey,
+                    key -> createAndInitDatasource(sqlDatasourceParams));
+
+        }
+        return sqlDatasourceToBeReturned;
+    }
+
+    private static SQLDatasource createAndInitDatasource(SQLDatasource.SQLDatasourceParams sqlDatasourceParams) {
+        SQLDatasource newSqlDatasource = new SQLDatasource(sqlDatasourceParams);
+        newSqlDatasource.incrementClientCounter();
+        return newSqlDatasource;
+    }
+
+
+    private Connection getSQLConnection() throws SQLException {
+        return hikariDataSource.getConnection();
+    }
+
+    private void closeConnectionPool() {
+        hikariDataSource.close();
+        poolShutdown = true;
+    }
+
+    private boolean isPoolShutdown() {
+        return poolShutdown;
+    }
+
+    private void incrementClientCounter() {
+        clientCounter.incrementAndGet();
+    }
+
+    public void decrementClientCounterAndAttemptPoolShutdown() {
+        acquireMutex();
+        if (!poolShutdown) {
+            if (clientCounter.decrementAndGet() == 0) {
+                closeConnectionPool();
+            }
+        }
+        releaseMutex();
+    }
+
+    private void releaseMutex() {
+        mutex.unlock();
+    }
+
+    private void acquireMutex() {
+        mutex.lock();
+    }
+
+    private void buildDataSource(SQLDatasourceParams sqlDatasourceParams) {
+        try {
+            HikariConfig config = new HikariConfig();
+            config.setJdbcUrl(sqlDatasourceParams.url);
+            config.setUsername(sqlDatasourceParams.user);
+            config.setPassword(sqlDatasourceParams.password);
+            if (sqlDatasourceParams.datasourceName != null && !sqlDatasourceParams.datasourceName.isEmpty()) {
+                if (sqlDatasourceParams.options == null || !sqlDatasourceParams.options
+                        .containsKey(Constants.Options.URL)) {
+                    //It is required to set the url to the datasource property when the
+                    //datasource class name is provided. Because according to hikari
+                    //either jdbcUrl or datasourceClassName will be honoured.
+                    config.addDataSourceProperty(Constants.Options.URL, sqlDatasourceParams.url);
+                }
+            }
+            config.setDataSourceClassName(sqlDatasourceParams.datasourceName);
+            if (sqlDatasourceParams.connectionPool != null) {
+                int maxOpenConn = sqlDatasourceParams.connectionPool.
+                        getIntValue(Constants.ConnectionPool.MAX_OPEN_CONNECTIONS).intValue();
+                if (maxOpenConn < 0) {
```

Comment:
Is this condition correct? I think it should be `maxOpenConn > 0`

Reply:
Yes, it should be greater than zero.